src/unit_test.c

```diff
@@ -0,0 +1,406 @@
+     * diagnostic information. Happens right at the start of main because
+     * setbuf must be used before any other operation on the stream. */
+    setbuf(stdout, NULL);
+    /* Also disable buffering for stderr because it's not guaranteed that it's
+     * unbuffered on all systems. */
+    setbuf(stderr, NULL);
+
+    /* Parse env args */
+    if (read_env() != 0) return EXIT_FAILURE;
+
+    /* Parse command-line args */
+    if (argc > 1) {
+        int named_arg_start = 1; /* index to begin processing named arguments */
+        if (argc > MAX_ARGS) {
+            fprintf(stderr, "Too many command-line arguments (max: %d)\n", MAX_ARGS);
+            return EXIT_FAILURE;
+        }
+
+        /* Check if we need to print help */
+        if (argv[1] && (strcmp(argv[1], "-help") == 0 || strcmp(argv[1], "-h") == 0)) {
+            help();
+            exit(EXIT_SUCCESS);
+        }
+
+        /* Check if we need to print the available tests */
+        if (argv[1] && strcmp(argv[1], "-print_tests") == 0) {
+            print_test_list(tf);
+            exit(EXIT_SUCCESS);
+        }
+
+        /* Compatibility Note: The first two args were the number of iterations and the seed. */
+        /* If provided, parse them and adjust the starting index for named arguments accordingly. */
+        if (argv[1] && argv[1][0] != '-') {
+            int has_seed = argc > 2 && argv[2] && argv[2][0] != '-';
+            if (parse_iterations(argv[1]) != 0) return EXIT_FAILURE;
+            if (has_seed) tf->args.custom_seed = (strcmp(argv[2], "NULL") == 0) ? NULL : argv[2];
+            named_arg_start = has_seed ? 3 : 2;
+        }
+        if (read_args(argc, argv, named_arg_start, tf) != 0) {
+            return EXIT_FAILURE;
+        }
+    }
+
+    tf->fn_run_test = tf->args.logging ? run_test_log : run_test;
+    return EXIT_SUCCESS;
+}
+
+static int tf_run(struct TestFramework* tf) {
+    /* Process exit status */
+    int status;
+    /* Whether to run all tests */
+    int run_all;
+    /* Loop iterator */
+    int it;
+    /* Initial test time */
+    int64_t start_time = gettime_i64(); /* maybe move this after the slots set */
+    /* Verify 'tf_init' has been called */
+    if (!tf->fn_run_test) {
+        fprintf(stderr, "Error: No test runner set. You must call 'tf_init' first to initialize the framework "
+                        "or manually assign 'fn_run_test' before calling 'tf_run'.\n");
+        return EXIT_FAILURE;
+    }
+
+    /* Populate targets with all tests if none were explicitly specified */
+    run_all = tf->args.targets.size == 0;
+    if (run_all) {
+        TestRef ref;
+        int slot = 0;
+        for (ref.group = 0; ref.group < tf->num_modules; ref.group++) {
+            struct TestModule* group = &tf->registry_modules[ref.group];
+            for (ref.idx = 0; ref.idx < group->size; ref.idx++) {
+                tf->args.targets.slots[slot++] = ref;
+                if (slot >= MAX_ARGS) {
```

Comment:
Same issue here with the `MAX_ARGS` check. I'd also move the guard to before the write. Sorry for the piecemeal comments!